presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -588,120 +590,111 @@ public PlanWithProperties visitTopNRowNumber(TopNRowNumberNode node, StreamPrefe
-                                    originalTableWriterNode.getNotNullColumnVariables(),
-                                    originalTableWriterNode.getTablePartitioningScheme(),
-                                    statisticAggregations.map(StatisticAggregations.Parts::getPartialAggregation),
-                                    originalTableWriterNode.getTaskCountIfScaledWriter(),
-                                    originalTableWriterNode.getIsTemporaryTableWriter()),
-                            fixedParallelism(),
-                            fixedParallelism());
+            if (nativeExecution && isNativeExecutionScaleWritersThreadsEnabled(session)) {
+                PlanWithProperties source = accept(tableWrite.getSource(), defaultParallelism(session));
+                PartitioningScheme partitioningScheme;
+                if (tableWrite.getTablePartitioningScheme().isPresent()) {
+                    partitioningScheme = tableWrite.getTablePartitioningScheme().get();
+                    verify(partitioningScheme.isScaleWriters());
                 }
                 else {
-                    PlanWithProperties source = accept(originalTableWriterNode.getSource(), defaultParallelism(session));
-                    PlanWithProperties exchange = deriveProperties(
-                            roundRobinExchange(idAllocator.getNextId(), LOCAL, source.getNode()),
-                            source.getProperties());
-                    tableWriter = deriveProperties(
-                            new TableWriterNode(
-                                    originalTableWriterNode.getSourceLocation(),
-                                    originalTableWriterNode.getId(),
-                                    originalTableWriterNode.getStatsEquivalentPlanNode(),
-                                    exchange.getNode(),
-                                    originalTableWriterNode.getTarget(),
-                                    variableAllocator.newVariable("partialrowcount", BIGINT),
-                                    variableAllocator.newVariable("partialfragments", VARBINARY),
-                                    variableAllocator.newVariable("partialcontext", VARBINARY),
-                                    originalTableWriterNode.getColumns(),
-                                    originalTableWriterNode.getColumnNames(),
-                                    originalTableWriterNode.getNotNullColumnVariables(),
-                                    originalTableWriterNode.getTablePartitioningScheme(),
-                                    statisticAggregations.map(StatisticAggregations.Parts::getPartialAggregation),
-                                    originalTableWriterNode.getTaskCountIfScaledWriter(),
-                                    originalTableWriterNode.getIsTemporaryTableWriter()),
-                            exchange.getProperties());
+                    partitioningScheme = new PartitioningScheme(
```

Comment:
if partitioningScheme is missing, does that mean that the table is unpartitioned? what does writer thread scaling do for us in that case?

Reply:
> if partitioningScheme is missing, does that mean that the table is unpartitioned?

Yes

> what does writer thread scaling do for us in that case?

There is an implementation for unpartitioned table writer. It is implemented as a local round robin exchange. The implementation starts with sending all data to a first consumer as as it gets more busy the implementation will start sending data to next one: https://github.com/facebookincubator/velox/blob/main/velox/exec/ScaleWriterLocalPartition.cpp#L274